File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.py
class Solution(object):
    def evaluate(self, s, knowledge):
        """
        :type s: str
        :type knowledge: List[List[str]]
        :rtype: str
        """
        mp = {}

        for item in knowledge:
            mp[item[0]] = item[1]

        key = ""
        res = ""
        flag = False

        for ch in s:
            if ch == '(':
                flag = True
            elif ch == ')':
                if key in mp:
                    res += mp[key]
                else:
                    res += "?"

                flag = False
                key = ""
            elif flag:
                key += ch
            else:
                res += ch

        return res

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        unordered_map<string, string> mp;

        for (auto& list : knowledge) {
            mp[list[0]] = list[1];
        }

        string key = "";
        string res = "";
        bool flag = false;

        for (char ch : s) {
            if (ch == '(') {
                flag = true;
            } else if (ch == ')') {
                if (mp.find(key) != mp.end()) {
                    res += mp[key];
                } else {
                    res += "?";
                }

                flag = false;
                key = "";
            } else if (flag) {
                key += ch;
            } else {
                res += ch;
            }
        }

        return res;
    }
};
